### PR TITLE
add Endpoint fields and endpoints to Group

### PR DESCRIPTION
**Pull request by kczimm**
- Added the TopicKind_t and ReliabilityKind_t fields to Endpoint.
- Added a vector of endpoints to Group based on my understanding of Figure 8.5 of the RTPS spec.

### CODE REVIEW

File: src/structure/endpoint.rs [hunk 1]
```diff
@@ -27,3 +29,18 @@ impl Entity for Endpoint {
         self.guid
     }
 }
+
+/// Used to indicate whether the [Endpoint] supports instance lifecycle
```
Comment (danieleades) [inline src/structure/endpoint.rs:33]: backticks around `endpoint`

File: src/structure/endpoint.rs [hunk 2]
```diff
@@ -27,3 +29,18 @@ impl Entity for Endpoint {
...
+/// management operations
+#[derive(Debug)]
+pub enum TopicKind {
+    WithKey,
+    NoKey,
+}
+
+/// The levels of reliability supported by the [Endpoint].
```
Comment (danieleades) [inline src/structure/endpoint.rs:41]: backticks around `endpoint`

File: src/structure/group.rs [hunk 3]
```diff
@@ -10,6 +10,7 @@ use super::{entity::Entity, guid::Guid};
 pub struct Group<T> {
     guid: Guid,
     marker: std::marker::PhantomData<T>,
+    endpoints: Vec<Endpoint>,
```
Comment (danieleades) [inline src/structure/group.rs:13]: I wasn't sure about this part, which I was I had left it out for now. Figure 8.5 certainly indicates that a `Group` should contain multiple `Endpoint`s, but I wasn't sure if this meant they should be stored directly in the group, or if the relationship was captured conceptually by each `Endpoint` having the same `Guid` prefix as the `Group` that contains them.
I was going to wait until I began implementing the behaviour/discovery modules to see which fell out naturally.

That said, I'm very open to being convinced!

File: src/structure/group.rs [hunk 4]
```diff
@@ -10,6 +10,7 @@ use super::{entity::Entity, guid::Guid};
 pub struct Group<T> {
     guid: Guid,
     marker: std::marker::PhantomData<T>,
```
Comment (danieleades) [inline src/structure/group.rs:12]: I wonder if the `Reader`/`Writer` marker type should be moved from the `Group` and into the `Endpoint`

**Review by danieleades — CHANGES_REQUESTED** [review]
thanks again for the contribution, really pleased to have you on board.

more questions than answers in this review i'm afraid

File: src/structure/endpoint.rs (hunk 1 above)
Comment (danieleades) [reply src/structure/endpoint.rs:33]: my rule of thumb is "when clippy::pedantic complains"

any references to objects in code should use backticks, which formats the object like `this`.
square brackets creates a link to the documentation for that object. so internal links should use both backticks and square brackets

File: src/structure/group.rs (hunk 3 above)
Comment (kczimm) [reply src/structure/group.rs:13]: > it's a very different beast.

Right, I see that. The `InnerParticipant` is definitely getting low-level. https://github.com/jhelovuo/RustDDS/blob/4c10adf0bc7ce4efd5096903e03954ba145332fb/src/dds/pubsub.rs#L369

However, this shouldn't be the equivalent type here. There is a DDS DomainParticipant and an RTPS Participant. They are one-to-one AFAIK but the DDS DomainParticipant is the user facing type. See page 21 of the spec:

> RTPS Participant is the container of RTPS Group entities which contain Endpoint entities. The RTPS 
Participant maps to a DDS DomainParticipant.

File: src/structure/group.rs (hunk 4 above)
Comment (danieleades) [reply src/structure/group.rs:12]: > I'm not sure the value of the `Group`

based on the 'structure' module alone, i'm inclined to agree with you. Perhaps the point of that extra layer of abstraction will become apparent in the 'behaviour' or 'discovery' modules...

then again maybe not. This, like many specs of this sort, was designed by committee. You always end up with strange design choices in that environment!

File: src/structure/group.rs (hunk 3 above)
Comment (jhelovuo) [reply src/structure/group.rs:13]: `Group` , as used in the RTPS spec, is an abstraction meaning "either a Publisher or a Subscriber".

AFAIK it has practical consequences only if you are implementing the QoS features "Group Ordered Access" or "Coherent Sets". As long as you do not support those, then grouping (e.g.) Readers under a Subscriber has little or no meaning. Group membership of Readers/Writers is not communicated over RTPS unless those QoS features are used.

All the Readers could be directly contained within a DomainParticipant with no intermediate Subscriber layer - and analogously for writing side. That is why RustDDS code has no concept of a Group.

**Pull request closed without merging**